File: backend/internal/services/job_service.go
package services

import (
	"context"
	"time"

	"microbridge/backend/internal/dto"
	"microbridge/backend/internal/models"
	"microbridge/backend/internal/repository"
	apperrors "microbridge/backend/internal/shared/errors"

	"github.com/google/uuid"
)

type JobService interface {
	CreateJob(ctx context.Context, employerID string, req dto.CreateJobRequest) (*dto.JobResponse, error)
	GetJob(ctx context.Context, id string) (*dto.JobResponse, error)
	UpdateJob(ctx context.Context, jobID string, employerID string, req dto.UpdateJobRequest) (*dto.JobResponse, error)
	DeleteJob(ctx context.Context, jobID string, employerID string) error
	ListJobs(ctx context.Context, filters dto.JobFilters, page, limit int) (*dto.PaginatedJobResponse, error)
	GetJobsByEmployer(ctx context.Context, employerID string, page, limit int) (*dto.PaginatedJobResponse, error)
	SearchJobs(ctx context.Context, query string, filters dto.JobFilters, page, limit int) (*dto.PaginatedJobResponse, error)
}

type jobService struct {
	jobRepo repository.JobRepository
}

func NewJobService(jobRepo repository.JobRepository) JobService {
	return &jobService{
		jobRepo: jobRepo,
	}
}

func (s *jobService) CreateJob(ctx context.Context, employerID string, req dto.CreateJobRequest) (*dto.JobResponse, error) {
	// Validate the request
	if err := s.validateCreateJobRequest(req); err != nil {
		return nil, err
	}

	// Create job model from request
	job := &models.Job{
		ID:           uuid.New().String(),
		EmployerID:   employerID,
		Title:        req.Title,
		Description:  req.Description,
		Company:      req.Company,
		Location:     req.Location,
		Duration:     req.Duration,
		Category:     req.Category,
		IsRemote:     req.IsRemote,
		JobType:      req.JobType,
		Status:       "draft", // Start as draft
		Skills:       s.convertToRequiredSkills(req.Skills),
		ExperienceLevel: req.ExperienceLevel,
		Salary:       req.Salary,
		Benefits:     req.Benefits,
		Requirements: req.Requirements,
		WorkArrangement: req.WorkArrangement,
		PreferredCandidates: req.PreferredCandidates,
		ApplicationDeadline: req.ApplicationDeadline,
		StartDate:    req.StartDate,
		EndDate:      req.EndDate,
		Views:        0,
		Applications: 0,
		CreatedAt:    time.Now(),
		UpdatedAt:    time.Now(),
	}

	if err := s.jobRepo.Create(ctx, job); err != nil {
		return nil, err
	}

	return s.jobToResponse(job), nil
}

func (s *jobService) GetJob(ctx context.Context, id string) (*dto.JobResponse, error) {
	job, err := s.jobRepo.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}

	return s.jobToResponse(job), nil
}

func (s *jobService) UpdateJob(ctx context.Context, jobID string, employerID string, req dto.UpdateJobRequest) (*dto.JobResponse, error) {
	job, err := s.jobRepo.GetByID(ctx, jobID)
	if err != nil {
		return nil, err
	}

	// Check if the employer owns this job
	if job.EmployerID != employerID {
		return nil, apperrors.NewAppError(403, "You don't have permission to update this job", nil)
	}

	// Update fields if provided
	if req.Title != nil {
		job.Title = *req.Title
	}
	if req.Description != nil {
		job.Description = *req.Description
	}
	if req.Location != nil {
		job.Location = *req.Location
	}
	if req.Category != nil {
		job.Category = *req.Category
	}
	if req.IsRemote != nil {
		job.IsRemote = *req.IsRemote
	}
	if req.JobType != nil {
		job.JobType = *req.JobType
	}
	if req.ExperienceLevel != nil {
		job.ExperienceLevel = *req.ExperienceLevel
	}
	if req.Status != nil {
		job.Status = *req.Status
	}
	if req.Duration != nil {
		job.Duration = *req.Duration
	}
	if req.Skills != nil && len(req.Skills) > 0 {
		job.Skills = s.convertToRequiredSkills(req.Skills)
	}
	if req.Salary != nil {
		job.Salary = *req.Salary
	}
	if req.Benefits != nil {
		job.Benefits = req.Benefits
	}
	if req.Requirements != nil {
		job.Requirements = req.Requirements
	}
	if req.WorkArrangement != nil {
		job.WorkArrangement = *req.WorkArrangement
	}
	if req.PreferredCandidates != nil {
		job.PreferredCandidates = *req.PreferredCandidates
	}
	if req.ApplicationDeadline != nil {
		job.ApplicationDeadline = req.ApplicationDeadline
	}
	if req.StartDate != nil {
		job.StartDate = req.StartDate
	}
	if req.EndDate != nil {
		job.EndDate = req.EndDate
	}

	job.UpdatedAt = time.Now()

	if err := s.jobRepo.Update(ctx, job); err != nil {
		return nil, err
	}

	return s.jobToResponse(job), nil
}

func (s *jobService) DeleteJob(ctx context.Context, jobID string, employerID string) error {
	job, err := s.jobRepo.GetByID(ctx, jobID)
	if err != nil {
		return err
	}

	// Check if the employer owns this job
	if job.EmployerID != employerID {
		return apperrors.NewAppError(403, "You don't have permission to delete this job", nil)
	}

	return s.jobRepo.Delete(ctx, jobID)
}

func (s *jobService) ListJobs(ctx context.Context, filters dto.JobFilters, page, limit int) (*dto.PaginatedJobResponse, error) {
	if page <= 0 {
		page = 1
	}
	if limit <= 0 || limit > 100 {
		limit = 20
	}

	offset := (page - 1) * limit
	filterMap := s.filtersToMap(filters)

	jobs, total, err := s.jobRepo.List(ctx, filterMap, limit, offset)
	if err != nil {
		return nil, err
	}

	jobResponses := make([]*dto.JobResponse, len(jobs))
	for i, job := range jobs {
		jobResponses[i] = s.jobToResponse(job)
	}

	return &dto.PaginatedJobResponse{
		Jobs: jobResponses,
		Pagination: dto.PaginationResponse{
			Page:     page,
			Limit:    limit,
			Total:    total,
			HasMore:  int64(page*limit) < total,
		},
	}, nil
}

func (s *jobService) GetJobsByEmployer(ctx context.Context, employerID string, page, limit int) (*dto.PaginatedJobResponse, error) {
	if page <= 0 {
		page = 1
	}
	if limit <= 0 || limit > 100 {
		limit = 20
	}

	offset := (page - 1) * limit

	jobs, total, err := s.jobRepo.GetByEmployerID(ctx, employerID, limit, offset)
	if err != nil {
		return nil, err
	}

	jobResponses := make([]*dto.JobResponse, len(jobs))
	for i, job := range jobs {
		jobResponses[i] = s.jobToResponse(job)
	}

	return &dto.PaginatedJobResponse{
		Jobs: jobResponses,
		Pagination: dto.PaginationResponse{
			Page:     page,
			Limit:    limit,
			Total:    total,
			HasMore:  int64(page*limit) < total,
		},
	}, nil
}

func (s *jobService) SearchJobs(ctx context.Context, query string, filters dto.JobFilters, page, limit int) (*dto.PaginatedJobResponse, error) {
	if page <= 0 {
		page = 1
	}
	if limit <= 0 || limit > 100 {
		limit = 20
	}

	offset := (page - 1) * limit
	filterMap := s.filtersToMap(filters)

	jobs, total, err := s.jobRepo.Search(ctx, query, filterMap, limit, offset)
	if err != nil {
		return nil, err
	}

	jobResponses := make([]*dto.JobResponse, len(jobs))
	for i, job := range jobs {
		jobResponses[i] = s.jobToResponse(job)
	}

	return &dto.PaginatedJobResponse{
		Jobs: jobResponses,
		Pagination: dto.PaginationResponse{
			Page:     page,
			Limit:    limit,
			Total:    total,
			HasMore:  int64(page*limit) < total,
		},
	}, nil
}

// Helper methods

func (s *jobService) validateCreateJobRequest(req dto.CreateJobRequest) error {
	if req.Title == "" {
		return apperrors.NewAppError(400, "Job title is required", nil)
	}
	if req.Description == "" {
		return apperrors.NewAppError(400, "Job description is required", nil)
	}
	if req.Category == "" {
		return apperrors.NewAppError(400, "Job category is required", nil)
	}
	if len(req.Skills) == 0 {
		return apperrors.NewAppError(400, "At least one skill is required", nil)
	}

	return nil
}

func (s *jobService) convertToRequiredSkills(skillNames []dto.SkillRequest) models.RequiredSkillsArray {
	skills := make(models.RequiredSkillsArray, len(skillNames))
	for i, skillReq := range skillNames {
		skills[i] = models.RequiredSkill{
			Name:       skillReq.Name,
			Level:      skillReq.Level,
			IsRequired: skillReq.IsRequired,
			Importance: skillReq.Importance,
			CanLearn:   skillReq.CanLearn,
		}
	}
	return skills
}

func (s *jobService) filtersToMap(filters dto.JobFilters) map[string]interface{} {
	filterMap := make(map[string]interface{})

	if filters.Status != "" {
		filterMap["status"] = filters.Status
	}
	if filters.Location != "" {
		filterMap["location"] = filters.Location
	}
	if filters.ExperienceLevel != "" {
		filterMap["experience_level"] = filters.ExperienceLevel
	}
	if filters.IsRemote != nil {
		filterMap["is_remote"] = *filters.IsRemote
	}
	if filters.EmployerID != "" {
		filterMap["employer_id"] = filters.EmployerID
	}
	if filters.Category != "" {
		filterMap["category"] = filters.Category
	}
	if filters.JobType != "" {
		filterMap["job_type"] = filters.JobType
	}

	return filterMap
}

func (s *jobService) jobToResponse(job *models.Job) *dto.JobResponse {
	return &dto.JobResponse{
		ID:              job.ID,
		EmployerID:      job.EmployerID,
		Title:           job.Title,
		Description:     job.Description,
		Company:         job.Company,
		CompanyID:       job.CompanyID,
		Skills:          s.convertFromRequiredSkills(job.Skills),
		ExperienceLevel: job.ExperienceLevel,
		Location:        job.Location,
		Duration:        job.Duration,
		Category:        job.Category,
		IsRemote:        job.IsRemote,
		JobType:         job.JobType,
		Salary:          job.Salary,
		Benefits:        job.Benefits,
		Requirements:    job.Requirements,
		WorkArrangement: job.WorkArrangement,
		ApplicationDeadline: job.ApplicationDeadline,
		StartDate:       job.StartDate,
		EndDate:         job.EndDate,
		PreferredCandidates: job.PreferredCandidates,
		Status:          job.Status,
		Views:           job.Views,
		Applications:    job.Applications,
		ReviewDueDate:   job.ReviewDueDate,
		CompletedAt:     job.CompletedAt,
		HiredStudentID:  job.HiredStudentID,
		CreatedAt:       job.CreatedAt,
		UpdatedAt:       job.UpdatedAt,
	}
}

func (s *jobService) convertFromRequiredSkills(skills models.RequiredSkillsArray) []dto.SkillResponse {
	skillResponses := make([]dto.SkillResponse, len(skills))
	for i, skill := range skills {
		skillResponses[i] = dto.SkillResponse{
			Name:       skill.Name,
			Level:      skill.Level,
			IsRequired: skill.IsRequired,
			Importance: skill.Importance,
			CanLearn:   skill.CanLearn,
		}
	}
	return skillResponses
}

File: frontend/src/services/jobService.ts
import { api } from './api';
import type { ApiResponse } from './authService';

export interface SkillRequest {
  name: string;
  level: number;
  is_required: boolean;
  importance: number;
  can_learn: boolean;
}

export interface SkillResponse {
  name: string;
  level: number;
  is_required: boolean;
  importance: number;
  can_learn: boolean;
}

export interface SalaryRange {
  currency: string;
  min_amount: number;
  max_amount: number;
  negotiable: boolean;
}

export interface WorkArrangement {
  schedule_type: string;
  hours_per_week: number;
  timezone: string;
  meeting_frequency: string;
}

export interface CandidatePreferences {
  preferred_locations: string[];
  min_experience_level: string;
  required_skills: string[];
  preferred_languages: string[];
  cultural_fit_requirements: string[];
}

export interface CreateJobRequest {
  title: string;
  description: string;
  company?: string;
  skills: SkillRequest[];
  experience_level: 'entry' | 'intermediate' | 'advanced' | 'senior' | 'expert';
  location?: string;
  duration?: number;
  category: string;
  is_remote: boolean;
  job_type: 'full-time' | 'part-time' | 'contract' | 'internship';
  salary?: SalaryRange;
  benefits?: string[];
  requirements?: string[];
  work_arrangement?: WorkArrangement;
  preferred_candidates?: CandidatePreferences;
  application_deadline?: string; // ISO date string
  start_date?: string; // ISO date string
  end_date?: string; // ISO date string
}

export interface UpdateJobRequest {
  title?: string;
  description?: string;
  skills?: SkillRequest[];
  experience_level?: 'entry' | 'intermediate' | 'advanced' | 'senior' | 'expert';
  location?: string;
  duration?: number;
  is_remote?: boolean;
  job_type?: 'full-time' | 'part-time' | 'contract' | 'internship';
  category?: string;
  status?: 'draft' | 'posted' | 'hired' | 'in_progress' | 'submitted' | 'review_pending' | 'completed' | 'disputed' | 'archived';
  salary?: SalaryRange;
  benefits?: string[];
  requirements?: string[];
  work_arrangement?: WorkArrangement;
  preferred_candidates?: CandidatePreferences;
  application_deadline?: string; // ISO date string
  start_date?: string; // ISO date string
  end_date?: string; // ISO date string
}

export interface JobFilters {
  category?: string;
  skills?: string[];
  location?: string;
  is_remote?: boolean;
  job_type?: string;
  experience_level?: string;
  status?: string;
  employer_id?: string;
}

export interface JobResponse {
  id: string;
  employer_id: string;
  title: string;
  description: string;
  company: string;
  company_id: string;
  skills: SkillResponse[];
  experience_level: string;
  location: string;
  duration: number;
  category: string;
  is_remote: boolean;
  job_type: string;
  salary: SalaryRange;
  benefits: string[];
  requirements: string[];
  work_arrangement: WorkArrangement;
  application_deadline?: string;
  start_date?: string;
  end_date?: string;
  preferred_candidates: CandidatePreferences;
  status: string;
  views: number;
  applications: number;
  review_due_date?: string;
  completed_at?: string;
  hired_student_id?: string;
  created_at: string;
  updated_at: string;
}

export interface PaginatedJobResponse {
  jobs: JobResponse[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    has_more: boolean;
  };
}

export interface JobRecommendation {
  job_id: string;
  title: string;
  description: string;
  skills: string[];
  budget: number;
  location: string;
  is_remote: boolean;
  match_score: number;
  reasons: string[];
  created_at: string;
}

class JobService {
  // Create a new job posting
  async createJob(jobData: CreateJobRequest): Promise<ApiResponse<JobResponse>> {
    try {
      const response = await api.post('/jobs', jobData);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to create job',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Get a job by ID
  async getJob(jobId: string): Promise<ApiResponse<JobResponse>> {
    try {
      const response = await api.get(`/jobs/${jobId}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to get job',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Update a job posting
  async updateJob(jobId: string, jobData: UpdateJobRequest): Promise<ApiResponse<JobResponse>> {
    try {
      const response = await api.put(`/jobs/${jobId}`, jobData);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to update job',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Delete a job posting
  async deleteJob(jobId: string): Promise<ApiResponse> {
    try {
      const response = await api.delete(`/jobs/${jobId}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to delete job',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // List jobs with filtering and pagination
  async listJobs(
    filters?: JobFilters,
    page: number = 1,
    limit: number = 20
  ): Promise<ApiResponse<PaginatedJobResponse>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        limit: limit.toString(),
        ...filters
      });

      // Handle boolean and array parameters
      if (filters?.is_remote !== undefined) {
        params.set('is_remote', filters.is_remote.toString());
      }
      if (filters?.skills && filters.skills.length > 0) {
        params.set('skills', filters.skills.join(','));
      }

      const response = await api.get(`/jobs?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to list jobs',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Search jobs
  async searchJobs(
    query: string,
    filters?: JobFilters,
    page: number = 1,
    limit: number = 20
  ): Promise<ApiResponse<PaginatedJobResponse>> {
    try {
      const params = new URLSearchParams({
        q: query,
        page: page.toString(),
        limit: limit.toString(),
        ...filters
      });

      // Handle boolean and array parameters
      if (filters?.is_remote !== undefined) {
        params.set('is_remote', filters.is_remote.toString());
      }
      if (filters?.skills && filters.skills.length > 0) {
        params.set('skills', filters.skills.join(','));
      }

      const response = await api.get(`/jobs/search?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to search jobs',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Get jobs by employer
  async getJobsByEmployer(
    employerId: string,
    page: number = 1,
    limit: number = 20
  ): Promise<ApiResponse<PaginatedJobResponse>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        limit: limit.toString()
      });

      const response = await api.get(`/jobs/employer/${employerId}?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to get jobs by employer',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Get job recommendations (from matching service)
  async getJobRecommendations(
    userId?: string,
    limit: number = 10
  ): Promise<ApiResponse<JobRecommendation[]>> {
    try {
      const params = new URLSearchParams({
        limit: limit.toString()
      });
      
      if (userId) {
        params.set('user_id', userId);
      }

      const response = await api.get(`/jobs/recommendations?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to get job recommendations',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Get jobs the current user has applied to
  async getAppliedJobs(
    page: number = 1,
    limit: number = 20
  ): Promise<ApiResponse<PaginatedJobResponse>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        limit: limit.toString()
      });

      const response = await api.get(`/jobs/applied?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to get applied jobs',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Get jobs posted by the current user (for employers)
  async getMyJobs(
    page: number = 1,
    limit: number = 20
  ): Promise<ApiResponse<PaginatedJobResponse>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        limit: limit.toString()
      });

      const response = await api.get(`/jobs/my-jobs?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to get my jobs',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }
}

export const jobService = new JobService();
export default jobService;